### PR TITLE
store/tikv: promote coverage for ticlient

### PR DESCRIPTION
**Pull request by wentaoxu**
as title says

### CODE REVIEW

**Comment by wentaoxu** [comment]
PTAL @zhangjinpeng1987 @disksing, only test for `txn`, I will add more for other module

File: util/arena/arena.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func NewAllocator(capacity int) *SimpleAllocator {
 // Alloc implements Allocator.AllocBytes interface.
 func (s *SimpleAllocator) Alloc(capacity int) []byte {
 	if s.off+capacity < cap(s.arena) {
-		slice := s.arena[s.off : s.off : s.off+capacity]
```
Comment (zimulala) [inline util/arena/arena.go]: Why change this？
It causes failure to pass the `gofmt`

Comment (jackysp) [reply util/arena/arena.go]: Please update golang to version 1.10, and try to make check again. @wentaoxu

File: store/tikv/2pc_fail_test.go [hunk 2]
```diff
@@ -100,3 +100,118 @@ func (s *testCommitterSuite) TestFailCommitTimeout(c *C) {
...
+	c.Assert(err, IsNil)
+	err = txn.Set([]byte("b"), []byte("b1"))
+	c.Assert(err, IsNil)
+
+	gofail.Enable("github.com/pingcap/tidb/store/tikv/tikvStoreSendReqResult", `1*return("timeout")`)
+	err = txn.Commit(context.Background())
+	c.Assert(err, NotNil)
+	gofail.Disable("github.com/pingcap/tidb/store/tikv/tikvStoreSendReqResult")
+
+	txnCheck := s.begin(c)
```
Comment (zhangjinpeng1987) [inline store/tikv/2pc_fail_test.go]: Also need to check `a`.

File: store/tikv/2pc_fail_test.go [hunk 3]
```diff
@@ -100,3 +100,118 @@ func (s *testCommitterSuite) TestFailCommitTimeout(c *C) {
...
+	err = txn.Set([]byte("b"), []byte("b1"))
+	c.Assert(err, IsNil)
+
+	gofail.Enable("github.com/pingcap/tidb/store/tikv/tikvStoreSendReqResult", `2*off->1*return("timeout")`)
+	err = txn.Commit(context.Background())
+	c.Assert(err, NotNil)
+	gofail.Disable("github.com/pingcap/tidb/store/tikv/tikvStoreSendReqResult")
+
+	txnCheck := s.begin(c)
+	var value []byte
```
Comment (zhangjinpeng1987) [inline store/tikv/2pc_fail_test.go]: ditto

**Comment by zhangjinpeng1987** [comment]
Any update?

File: util/arena/arena.go (hunk 1 above)
Comment (wentaoxu) [reply util/arena/arena.go]: ok

File: store/tikv/2pc_fail_test.go (hunk 2 above)
Comment (wentaoxu) [reply store/tikv/2pc_fail_test.go]: ok

File: store/tikv/2pc_fail_test.go (hunk 3 above)
Comment (wentaoxu) [reply store/tikv/2pc_fail_test.go]: ok

**Comment by wentaoxu** [comment]
2pc.go: 80.6% -> 86.5%
lock_resolve.go: 48% -> 64.8%
txn.go: 78.9% -> 94.4%

File: store/tikv/2pc_fail_test.go [hunk 4]
```diff
@@ -100,3 +116,300 @@ func (s *testCommitterSuite) TestFailCommitTimeout(c *C) {
...
+	c.Assert(err, IsNil)
+	c.Assert(value, BytesEquals, []byte("a1"))
+}
+
+func (s *testCommitterSuite) TestTwoPhaseCommitActionString(c *C) {
+	action := twoPhaseCommitAction(0)
+	c.Assert(action.String(), Equals, "unknown")
+}
+
+/*func (s *testCommitterSuite) TestFailNewTwoPhaseCommitter(c *C) {
```
Comment (zhangjinpeng1987) [inline store/tikv/2pc_fail_test.go]: Why comment this function?

File: store/tikv/lock_resolver.go [hunk 5]
```diff
@@ -15,10 +15,11 @@ package tikv
 
 import (
 	"container/list"
-	"fmt"
+	//"fmt"
```
Comment (zhangjinpeng1987) [inline store/tikv/lock_resolver.go]: ???

File: store/tikv/2pc_fail_test.go (hunk 4 above)
Comment (wentaoxu) [reply store/tikv/2pc_fail_test.go]: this case will do harm to other test cases. I will enable it in the future.

File: store/tikv/lock_resolver.go (hunk 5 above)
Comment (wentaoxu) [reply store/tikv/lock_resolver.go]: I will remote it.

File: store/tikv/2pc_fail_test.go [hunk 6]
```diff
@@ -19,6 +19,9 @@ import (
 	. "github.com/pingcap/check"
 	"github.com/pingcap/tidb/terror"
 	"golang.org/x/net/context"
+	//"github.com/pingcap/tidb/kv"
```
Comment (tiancaiamao) [inline store/tikv/2pc_fail_test.go]: clean up

File: store/tikv/2pc_fail_test.go (hunk 4 above)
Comment (tiancaiamao) [reply store/tikv/2pc_fail_test.go]: So add comment for it and use
```
c.Skip("reason")
```

File: store/tikv/txn_test.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
+package tikv
+
+import (
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/testleak"
+)
+
+type testTxnSuite struct {
```
Comment (tiancaiamao) [inline store/tikv/txn_test.go]: Inherit `oneByOneSuite`, otherwise integrated test with TiKV may fail.

File: store/tikv/txn_test.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(txn.String(), Equals, "1")
+}
+
+func (s *testTxnSuite) TestSeekReverse(c *C) {
+	defer testleak.AfterTest(c)()
+
+	txn, err := newTiKVTxn(s.store)
+	c.Assert(err, IsNil)
+	iter, err := txn.SeekReverse(nil)
```
Comment (tiancaiamao) [inline store/tikv/txn_test.go]: It test nothing

File: store/tikv/txn_test.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		v := values[i]
+		c.Assert(iter.Valid(), IsTrue)
+		c.Assert([]byte(iter.Key()), BytesEquals, k)
+		c.Assert(iter.Value(), BytesEquals, v)
+		c.Assert(iter.Next(), IsNil)
+	}
+	c.Assert(iter.Valid(), IsFalse)
+}
+
+func (s *testTxnSuite) TestSetAbdDelOption(c *C) {
```
Comment (tiancaiamao) [inline store/tikv/txn_test.go]: s/TestSetAbdDelOption/TestSetAndDelOption

File: store/tikv/txn_test.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+func (s *testTxnSuite) TestRollback(c *C) {
+	txn, err := newTiKVTxn(s.store)
+	c.Assert(err, IsNil)
+	c.Assert(txn.Set([]byte("a"), []byte("a1")), IsNil)
+	c.Assert(txn.Set([]byte("b"), []byte("b1")), IsNil)
+	c.Assert(txn.Set([]byte("c"), []byte("c1")), IsNil)
+	c.Assert(txn.Len(), Equals, 3)
+	c.Assert(txn.Rollback(), IsNil)
+	c.Assert(txn.Rollback(), NotNil)
```
Comment (tiancaiamao) [inline store/tikv/txn_test.go]: Maybe this behavior is undefined?

File: store/tikv/2pc_fail_test.go (hunk 4 above)
Comment (wentaoxu) [reply store/tikv/2pc_fail_test.go]: ok

File: store/tikv/2pc_fail_test.go (hunk 6 above)
Comment (wentaoxu) [reply store/tikv/2pc_fail_test.go]: ok

File: store/tikv/txn_test.go (hunk 7 above)
Comment (wentaoxu) [reply store/tikv/txn_test.go]: ok

File: store/tikv/txn_test.go (hunk 8 above)
Comment (wentaoxu) [reply store/tikv/txn_test.go]: It test whether it will report error if the input is nil.

File: store/tikv/txn_test.go (hunk 9 above)
Comment (wentaoxu) [reply store/tikv/txn_test.go]: ok

File: store/tikv/txn_test.go (hunk 10 above)
Comment (wentaoxu) [reply store/tikv/txn_test.go]: It will return `ErrInvalidTxn`, I think it is right to report error here.

**Comment by tiancaiamao** [comment]
LGTM
/run-all-tests

**Comment by shenli** [comment]
Close this PR due to out of date.

**Pull request closed without merging**